components/Banner/examples/supplemental/MDCBannerView.m

```diff
@@ -0,0 +1,190 @@
+      iconImageView.clipsToBounds = YES;
+      [iconImageViewContainer addSubview:iconImageView];
+      self.iconImageViewContainer = iconImageViewContainer;
+      self.iconImageView = iconImageView;
+      [self addSubview:iconImageViewContainer];
+    } else {
+      self.iconImageView.image = image;
+    }
+  } else {
+    _image = nil;
+    [self.iconImageViewContainer removeFromSuperview];
+    self.iconImageViewContainer = nil;
+  }
+  [self.layout reloadData];
+  [self setNeedsLayout];
+}
+
+- (void)setButtons:(NSArray<__kindof UIButton *> *)buttons {
+  for (UIButton *button in _buttons) {
+    [button removeFromSuperview];
+  }
+  for (UIButton *button in buttons) {
+    [self addSubview:button];
+  }
+  _buttons = buttons;
+  [self.layout reloadData];
+  [self setNeedsLayout];
+}
+
+#pragma mark - Banner Views
+
++ (MDCBannerView *)bannerWithText:(NSString *)text
+                            image:(UIImage *)image
+                          buttons:(NSArray<__kindof UIButton *> *)buttons {
+  MDCBannerView *mdcBannerView = [[MDCBannerView alloc] initWithFrame:CGRectZero];
+  mdcBannerView.text = text;
+  mdcBannerView.image = image;
+  mdcBannerView.buttons = buttons;
+  return mdcBannerView;
+}
+
+#pragma mark - UIView overrides
+
+- (CGSize)sizeThatFits:(CGSize)size {
+  self.layout = [[MDCBannerViewLayout alloc] initWithSizeToFit:size];
+  self.layout.dataSource = self;
+  [self.layout reloadData];
+  return [self.layout frameSize];
+}
+
+- (void)layoutSubviews {
+  [super layoutSubviews];
+
+  [self applyMDCBannerViewLayout:self.layout];
+}
+
+#pragma mark - MDCBannerViewLayoutDataSource
+
+- (UILabel *)textLabelForBannerViewLayout:(MDCBannerViewLayout *)bannerViewLayout {
+  return self.textLabel;
+}
+
+- (UIView *)imageContainerForBannerViewLayout:(MDCBannerViewLayout *)bannerViewLayout {
+  return self.iconImageViewContainer;
+}
+
+- (NSInteger)numberOfButtonsForBannerViewLayout:(MDCBannerViewLayout *)bannerViewLayout {
+  return self.buttons.count;
+}
+
+- (UIButton *)bannerViewLayout:(MDCBannerViewLayout *)bannerViewLayout
+                 buttonAtIndex:(NSInteger)index {
+  return self.buttons[index];
+}
+
+#pragma mark - Layout methods
+
+- (void)applyMDCBannerViewLayout:(MDCBannerViewLayout *)layout {
+  if (![self isMDCBannerViewLayoutApplicable:layout]) {
+    return;
+  }
+
+  self.iconImageViewContainer.frame = layout.imageContainerFrame;
+  self.iconImageView.center = CGPointMake(CGRectGetWidth(layout.imageContainerFrame) / 2,
+                                          CGRectGetHeight(layout.imageContainerFrame) / 2);
+  self.textLabel.frame = layout.textLabelFrame;
+  [self.buttons enumerateObjectsUsingBlock:^(id object, NSUInteger idx, BOOL *stop) {
+    if (idx >= layout.buttonFrames.count) {
+      *stop = YES;
+      return;
+    }
+    UIButton *button = (UIButton *)object;
+    CGRect buttonFrame = [layout.buttonFrames[idx] CGRectValue];
+    button.frame = buttonFrame;
+  }];
```

Comment:
I'm surprised we're doing layout here in this view if we have a layout object. Can we move this logic into the layout object as well?

Reply:
Followed up offline. This method will be kept.